### PR TITLE
fix: construction of contract call byte array args

### PR DESCRIPTION
**Pull request by gqln**
### Summary
This PR adds support for parsing fixed-size `byte` array arguments (e.g. `bytes32`) for contract calls.
This PR also modifies the parsing logic to expect a `0x`-prefixed hexadecimal representation of a byte array to be provided for all `bytes` arguments.

### Changes
**Prior Behavior**
All args with types prefixed with "bytes" were copied into byte slices.
For types like `bytes32`, geth would throw `abi: cannot use array as type slice as argument`.

Hexadecimal strings like "0xabcdef" were directly copied, leading to slices like `[]byte(["0", "x", "a", "b", ...])`.

**New Behavior**
`encodeMethodArgsStrings ` distinguishes between dynamically-sized byte array arguments and fixed-size byte array arguments.

`bytes` are still converted to byte slices, as is [required by geth](https://github.com/ethereum/go-ethereum/blob/master/accounts/abi/type.go#L250-L251).

Fixed-size arrays like `bytes32` are converted into golang byte arrays of the corresponding size, [as required by geth](https://github.com/ethereum/go-ethereum/blob/master/accounts/abi/type.go#L248-L249).

Hexadecimal strings like "0xabcdef" are trimmed and hex-decoded, resulting in `[0xab, 0xcd, 0xef]`.

If decoding fails, an error is returned.
If an invalid size (`bytes45`) is passed, an error is returned.
If the decoded byte array length doesn't match the expected size for fixed-size arrays, an error is returned.

### CODE REVIEW

File: services/construction_service.go [hunk 1]
```diff
@@ -1048,10 +1049,37 @@ func encodeMethodArgsStrings(sigData []byte, methodSig string, methodArgs []stri
...
+				if len(bytes) != size {
+					return nil, fmt.Errorf(
+						"received %d bytes for argument of type %s; expected %d bytes",
+						len(bytes), v, size,
+					)
+				}
+
+				arrayType := reflect.ArrayOf(size, reflect.TypeOf(byte(0)))
+				arrayValue := reflect.New(arrayType).Elem()
+				for i := 0; i < arrayValue.Len() && i < len(bytes); i++ {
```
Comment (jiangchuan-he-cb) [inline services/construction_service.go:1079]: I think we're safe to just put `i < len(bytes)` in the For statement to make it more clean, given `arrayValue` and `bytes` both size are the same, correct ?

**Review by jiangchuan-he-cb — APPROVED** [review]

**Review by GeekArthur — APPROVED** [review]

File: services/construction_service.go [hunk 2]
```diff
@@ -1048,10 +1049,44 @@ func encodeMethodArgsStrings(sigData []byte, methodSig string, methodArgs []stri
...
+				if err != nil {
+					return nil, err
+				}
 				argData = value
+			} else {
+				// Converts fixed-size byte array (like bytes32) to array
+				sizeStr := strings.TrimPrefix(v, "bytes")
+				size, err := strconv.Atoi(sizeStr)
+				if err != nil {
+					log.Fatal(err)
```
Comment (Inphi) [inline services/construction_service.go:1064]: Since we're handling user input, should return an error here rather than panic.

**Review by Inphi — COMMENTED** [review]
Looks good overall. Just have one comment on logging.

File: services/construction_service.go (hunk 2 above)
Comment (gqln) [reply services/construction_service.go]: fixed!

**Review by Inphi — APPROVED** [review]
Thanks!

**Pull request merged**